backend/src/xfd_django/xfd_api/auth_saml.py

```diff
@@ -0,0 +1,362 @@
+import os
+from typing import Any, Dict, Optional
+import urllib.parse
+
+# Third-Party Libraries
+from django.conf import settings as dj_settings
+from fastapi import APIRouter, HTTPException, Request
+from fastapi.responses import RedirectResponse, Response
+from onelogin.saml2.auth import OneLogin_Saml2_Auth
+from onelogin.saml2.idp_metadata_parser import OneLogin_Saml2_IdPMetadataParser
+from onelogin.saml2.settings import OneLogin_Saml2_Settings
+from xfd_mini_dl.models import User
+
+from .auth import (
+    create_jwt_token,
+    update_login_block_status,
+    user_to_dict,
+    validate_json_serialization,
+)
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+router = APIRouter()
+
+
+# =============================================================================
+# Env helpers & config
+# =============================================================================
+def _env_truthy(in_var: Optional[str]) -> bool:
+    """Return True if an environment variable-like string is truthy."""
+    if in_var is None:
+        return False
+    return in_var.strip().lower() in {"1", "true", "yes", "y", "on"}
+
+
+BACKEND_DOMAIN = (
+    os.getenv("BACKEND_DOMAIN") or os.getenv("APP_BASE_URL") or ""
+).rstrip("/")
+FRONTEND_DOMAIN = (
+    os.getenv("FRONTEND_DOMAIN") or os.getenv("FRONTEND_BASE_URL") or ""
+).rstrip("/")
+
+# Sourced from environment Django settings
+OKTA_METADATA_URL = os.getenv("OKTA_SAML_METADATA_URL")
+
+IS_LOCAL = _env_truthy(os.getenv("IS_LOCAL"))
+SAML_SP_CERT_PATH = os.getenv("SAML_SP_CERT_PATH")
+SAML_SP_PRIVATE_KEY_PATH = os.getenv("SAML_SP_PRIVATE_KEY_PATH")
+WANT_NAMEID_ENCRYPTED = _env_truthy(os.getenv("WANT_NAMEID_ENCRYPTED"))  # optional
+
+
+def _read_text_file(path: str) -> str:
+    """Read and return the contents of a text file."""
+    with open(path, encoding="utf-8") as f:
+        return f.read().strip()
+
+
+# =============================================================================
+# Injectable IdP parser + lazy settings cache to avoid import-time fetch errors
+# =============================================================================
+
+
+class _SamlConfig:
+    """Mutable SAML settings holder used to avoid module-level globals."""
+
+    idp_parser = OneLogin_Saml2_IdPMetadataParser
+    settings_cache: Optional[Dict[str, Any]] = None
+
+
+def reset_saml_settings_cache_for_tests() -> None:
+    """Clear cached SAML settings for tests."""
+    _SamlConfig.settings_cache = None
+
+
+def set_idp_metadata_parser_for_tests(parser_cls) -> None:
+    """Inject a fake IdP metadata parser for tests."""
+    _SamlConfig.idp_parser = parser_cls
+
+
+def _build_sp_settings() -> Dict[str, Any]:
+    """Build SAML settings dict by merging SP config with Okta IdP metadata."""
+    if not OKTA_METADATA_URL:
+        raise RuntimeError("OKTA_SAML_METADATA_URL is not set")
+
+    # Fetch & parse IdP metadata
+    idp_data = _SamlConfig.idp_parser.parse_remote(OKTA_METADATA_URL)
```

Comment:
Validate the URL before passing it into parse_remote. Wrap parse_remote in a try/except and log a clear error, then raise a controlled RuntimeError with a generic message, not the full stack trace or URL.

Reply:
CRASM-3481 was created to track the work for the URL validation. This will be addressed in another ticket.